File: .UNSAFE-ALL-BINDINGS-DELETED-FINAL-20260125/bindings/clojure/src/proven/safe_currency.clj
;; SPDX-License-Identifier: PMPL-1.0
;; SPDX-FileCopyrightText: 2025 Hyperpolymath

(ns proven.safe-currency
  "Safe currency operations with type-safe monetary values."
  (:require [clojure.string :as str]))

;; ISO 4217 currency codes with metadata
(def currency-codes
  "Set of supported ISO 4217 currency codes."
  #{:USD :EUR :GBP :JPY :CHF :CAD :AUD :NZD :CNY :INR
    :BRL :MXN :KRW :SGD :HKD :SEK :NOK :DKK :PLN :RUB
    :ZAR :TRY :THB :MYR :IDR :PHP :VND :AED :SAR :ILS
    :CZK :HUF :RON :BGN :HRK :ISK :CLP :COP :PEN :ARS
    :BTC :ETH})

(def currency-metadata
  "Metadata for currency codes including decimals, symbol, and name."
  {:USD {:decimals 2 :symbol "$" :name "US Dollar"}
   :EUR {:decimals 2 :symbol "€" :name "Euro"}
   :GBP {:decimals 2 :symbol "£" :name "British Pound"}
   :JPY {:decimals 0 :symbol "¥" :name "Japanese Yen"}
   :CHF {:decimals 2 :symbol "Fr" :name "Swiss Franc"}
   :CAD {:decimals 2 :symbol "C$" :name "Canadian Dollar"}
   :AUD {:decimals 2 :symbol "A$" :name "Australian Dollar"}
   :NZD {:decimals 2 :symbol "NZ$" :name "New Zealand Dollar"}
   :CNY {:decimals 2 :symbol "¥" :name "Chinese Yuan"}
   :INR {:decimals 2 :symbol "₹" :name "Indian Rupee"}
   :BRL {:decimals 2 :symbol "R$" :name "Brazilian Real"}
   :MXN {:decimals 2 :symbol "Mex$" :name "Mexican Peso"}
   :KRW {:decimals 0 :symbol "₩" :name "South Korean Won"}
   :SGD {:decimals 2 :symbol "S$" :name "Singapore Dollar"}
   :HKD {:decimals 2 :symbol "HK$" :name "Hong Kong Dollar"}
   :SEK {:decimals 2 :symbol "kr" :name "Swedish Krona"}
   :NOK {:decimals 2 :symbol "kr" :name "Norwegian Krone"}
   :DKK {:decimals 2 :symbol "kr" :name "Danish Krone"}
   :PLN {:decimals 2 :symbol "zł" :name "Polish Zloty"}
   :RUB {:decimals 2 :symbol "₽" :name "Russian Ruble"}
   :ZAR {:decimals 2 :symbol "R" :name "South African Rand"}
   :TRY {:decimals 2 :symbol "₺" :name "Turkish Lira"}
   :THB {:decimals 2 :symbol "฿" :name "Thai Baht"}
   :MYR {:decimals 2 :symbol "RM" :name "Malaysian Ringgit"}
   :IDR {:decimals 2 :symbol "Rp" :name "Indonesian Rupiah"}
   :PHP {:decimals 2 :symbol "₱" :name "Philippine Peso"}
   :VND {:decimals 0 :symbol "₫" :name "Vietnamese Dong"}
   :AED {:decimals 2 :symbol "د.إ" :name "UAE Dirham"}
   :SAR {:decimals 2 :symbol "﷼" :name "Saudi Riyal"}
   :ILS {:decimals 2 :symbol "₪" :name "Israeli Shekel"}
   :CZK {:decimals 2 :symbol "Kč" :name "Czech Koruna"}
   :HUF {:decimals 2 :symbol "Ft" :name "Hungarian Forint"}
   :RON {:decimals 2 :symbol "lei" :name "Romanian Leu"}
   :BGN {:decimals 2 :symbol "лв" :name "Bulgarian Lev"}
   :HRK {:decimals 2 :symbol "kn" :name "Croatian Kuna"}
   :ISK {:decimals 0 :symbol "kr" :name "Icelandic Krona"}
   :CLP {:decimals 0 :symbol "$" :name "Chilean Peso"}
   :COP {:decimals 2 :symbol "$" :name "Colombian Peso"}
   :PEN {:decimals 2 :symbol "S/" :name "Peruvian Sol"}
   :ARS {:decimals 2 :symbol "$" :name "Argentine Peso"}
   :BTC {:decimals 8 :symbol "₿" :name "Bitcoin"}
   :ETH {:decimals 8 :symbol "Ξ" :name "Ethereum"}})

(defrecord Money [minor-units currency])

(defn get-decimals
  "Get the number of decimal places for a currency."
  [currency-code]
  (get-in currency-metadata [currency-code :decimals] 2))

(defn get-symbol
  "Get the symbol for a currency."
  [currency-code]
  (get-in currency-metadata [currency-code :symbol] ""))

(defn get-currency-name
  "Get the full name of a currency."
  [currency-code]
  (get-in currency-metadata [currency-code :name] "Unknown Currency"))

(defn valid-code?
  "Check if a currency code is valid."
  [currency-code]
  (contains? currency-codes currency-code))

(defn parse-code
  "Parse a currency code from string.
   Returns {:ok keyword} or {:error message}."
  [code-string]
  (if (str/blank? code-string)
    {:error "empty_currency_code"}
    (let [currency-keyword (keyword (str/upper-case (str/trim code-string)))]
      (if (valid-code? currency-keyword)
        {:ok currency-keyword}
        {:error "unknown_currency_code"}))))

(defn parse-code!
  "Parse a currency code from string. Throws ex-info on error."
  [code-string]
  (let [result (parse-code code-string)]
    (if (:ok result)
      (:ok result)
      (throw (ex-info (str "Invalid currency code: " (:error result))
                      {:type :invalid-currency-code
                       :error (:error result)
                       :input code-string})))))

(defn- multiplier
  "Get the multiplier for converting major to minor units."
  [currency-code]
  (long (Math/pow 10 (get-decimals currency-code))))

(defn from-major
  "Create Money from major units (e.g., dollars, euros)."
  [amount currency-code]
  (when (valid-code? currency-code)
    (->Money (* (long amount) (multiplier currency-code)) currency-code)))

(defn from-minor
  "Create Money from minor units (e.g., cents, satoshis)."
  [amount currency-code]
  (when (valid-code? currency-code)
    (->Money (long amount) currency-code)))

(defn zero
  "Create zero Money for a currency."
  [currency-code]
  (from-minor 0 currency-code))

(defn get-currency
  "Get the currency code of a Money value."
  [money]
  (:currency money))

(defn get-minor
  "Get the amount in minor units."
  [money]
  (:minor-units money))

(defn get-major
  "Get the amount in major units (truncated)."
  [money]
  (quot (:minor-units money) (multiplier (:currency money))))

(defn get-major-decimal
  "Get the amount in major units as a BigDecimal."
  [money]
  (let [decimal-places (get-decimals (:currency money))]
    (bigdec (/ (bigdec (:minor-units money))
               (bigdec (multiplier (:currency money)))))))

(defn add
  "Add two Money values. Returns {:ok Money} or {:error message}."
  [money-a money-b]
  (if (not= (:currency money-a) (:currency money-b))
    {:error "currency_mismatch"}
    {:ok (->Money (+ (:minor-units money-a) (:minor-units money-b))
                  (:currency money-a))}))

(defn add!
  "Add two Money values. Throws ex-info on currency mismatch."
  [money-a money-b]
  (let [result (add money-a money-b)]
    (if (:ok result)
      (:ok result)
      (throw (ex-info "Currency mismatch"
                      {:type :currency-mismatch
                       :currency-a (:currency money-a)
                       :currency-b (:currency money-b)})))))

(defn sub
  "Subtract two Money values. Returns {:ok Money} or {:error message}."
  [money-a money-b]
  (if (not= (:currency money-a) (:currency money-b))
    {:error "currency_mismatch"}
    {:ok (->Money (- (:minor-units money-a) (:minor-units money-b))
                  (:currency money-a))}))

(defn sub!
  "Subtract two Money values. Throws ex-info on currency mismatch."
  [money-a money-b]
  (let [result (sub money-a money-b)]
    (if (:ok result)
      (:ok result)
      (throw (ex-info "Currency mismatch"
                      {:type :currency-mismatch
                       :currency-a (:currency money-a)
                       :currency-b (:currency money-b)})))))

(defn mul
  "Multiply Money by a scalar."
  [money scalar]
  (->Money (* (:minor-units money) (long scalar))
           (:currency money)))

(defn div
  "Divide Money by a scalar. Returns {:ok Money} or {:error message}."
  [money scalar]
  (if (zero? scalar)
    {:error "division_by_zero"}
    {:ok (->Money (quot (:minor-units money) (long scalar))
                  (:currency money))}))

(defn div!
  "Divide Money by a scalar. Throws ex-info on division by zero."
  [money scalar]
  (let [result (div money scalar)]
    (if (:ok result)
      (:ok result)
      (throw (ex-info "Division by zero"
                      {:type :division-by-zero})))))

(defn zero?-money
  "Check if Money is zero."
  [money]
  (zero? (:minor-units money)))

(defn positive?
  "Check if Money is positive."
  [money]
  (pos? (:minor-units money)))

(defn negative?
  "Check if Money is negative."
  [money]
  (neg? (:minor-units money)))

(defn abs
  "Get absolute value of Money."
  [money]
  (->Money (Math/abs ^long (:minor-units money))
           (:currency money)))

(defn negate
  "Negate a Money value."
  [money]
  (->Money (- (:minor-units money))
           (:currency money)))

(defn compare-money
  "Compare two Money values. Returns -1, 0, or 1.
   Returns {:ok result} or {:error message}."
  [money-a money-b]
  (if (not= (:currency money-a) (:currency money-b))
    {:error "currency_mismatch"}
    {:ok (compare (:minor-units money-a) (:minor-units money-b))}))

(defn format-money
  "Format Money for display."
  [money]
  (let [currency-code (:currency money)
        decimal-places (get-decimals currency-code)
        divisor (multiplier currency-code)
        abs-units (Math/abs ^long (:minor-units money))
        major-part (quot abs-units divisor)
        minor-part (rem abs-units divisor)
        sign-prefix (if (neg? (:minor-units money)) "-" "")
        currency-symbol (get-symbol currency-code)]
    (if (zero? decimal-places)
      (str sign-prefix currency-symbol major-part)
      (str sign-prefix currency-symbol major-part "."
           (format (str "%0" decimal-places "d") minor-part)))))

(defn format-iso
  "Format Money in ISO format (amount currency-code)."
  [money]
  (let [decimal-places (get-decimals (:currency money))
        divisor (multiplier (:currency money))
        amount (/ (double (:minor-units money)) (double divisor))]
    (str (format (str "%." decimal-places "f") amount) " " (name (:currency money)))))

(defn allocate
  "Allocate Money according to ratios. Returns vector of Money values."
  [money ratios]
  (let [total-ratio (reduce + ratios)
        base-amount (quot (:minor-units money) total-ratio)
        remainder (rem (:minor-units money) total-ratio)
        allocations (mapv #(* base-amount %) ratios)]
    ;; Distribute remainder to first n allocations
    (mapv (fn [allocation-amount idx]
            (->Money (+ allocation-amount (if (< idx remainder) 1 0))
                     (:currency money)))
          allocations
          (range))))

(defn split
  "Split Money equally into n parts. Returns vector of Money values."
  [money num-parts]
  (when (pos? num-parts)
    (let [base-amount (quot (:minor-units money) num-parts)
          remainder (rem (:minor-units money) num-parts)]
      (mapv (fn [idx]
              (->Money (+ base-amount (if (< idx remainder) 1 0))
                       (:currency money)))
            (range num-parts)))))

File: .UNSAFE-ALL-BINDINGS-DELETED-FINAL-20260125/bindings/php/src/Proven/SafeCurrency.php
<?php
// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2025 Hyperpolymath

declare(strict_types=1);

namespace Proven;

/**
 * ISO 4217 currency codes with minor unit information.
 *
 * This enum includes the most commonly used currencies. The minorUnits
 * value indicates the number of decimal places for the currency.
 */
enum CurrencyCode: string
{
    // Major world currencies
    case USD = 'USD';  // United States Dollar
    case EUR = 'EUR';  // Euro
    case GBP = 'GBP';  // British Pound Sterling
    case JPY = 'JPY';  // Japanese Yen
    case CHF = 'CHF';  // Swiss Franc
    case CNY = 'CNY';  // Chinese Yuan Renminbi
    case AUD = 'AUD';  // Australian Dollar
    case CAD = 'CAD';  // Canadian Dollar
    case NZD = 'NZD';  // New Zealand Dollar
    case HKD = 'HKD';  // Hong Kong Dollar
    case SGD = 'SGD';  // Singapore Dollar

    // European currencies
    case SEK = 'SEK';  // Swedish Krona
    case NOK = 'NOK';  // Norwegian Krone
    case DKK = 'DKK';  // Danish Krone
    case PLN = 'PLN';  // Polish Zloty
    case CZK = 'CZK';  // Czech Koruna
    case HUF = 'HUF';  // Hungarian Forint
    case RON = 'RON';  // Romanian Leu
    case BGN = 'BGN';  // Bulgarian Lev
    case ISK = 'ISK';  // Icelandic Krona
    case RUB = 'RUB';  // Russian Ruble
    case UAH = 'UAH';  // Ukrainian Hryvnia
    case TRY = 'TRY';  // Turkish Lira

    // Asian currencies
    case KRW = 'KRW';  // South Korean Won
    case INR = 'INR';  // Indian Rupee
    case IDR = 'IDR';  // Indonesian Rupiah
    case MYR = 'MYR';  // Malaysian Ringgit
    case THB = 'THB';  // Thai Baht
    case PHP = 'PHP';  // Philippine Peso
    case VND = 'VND';  // Vietnamese Dong
    case TWD = 'TWD';  // Taiwan Dollar

    // Middle Eastern currencies
    case AED = 'AED';  // UAE Dirham
    case SAR = 'SAR';  // Saudi Riyal
    case ILS = 'ILS';  // Israeli New Shekel
    case QAR = 'QAR';  // Qatari Riyal
    case KWD = 'KWD';  // Kuwaiti Dinar
    case BHD = 'BHD';  // Bahraini Dinar
    case OMR = 'OMR';  // Omani Rial
    case JOD = 'JOD';  // Jordanian Dinar

    // American currencies
    case MXN = 'MXN';  // Mexican Peso
    case BRL = 'BRL';  // Brazilian Real
    case ARS = 'ARS';  // Argentine Peso
    case CLP = 'CLP';  // Chilean Peso
    case COP = 'COP';  // Colombian Peso
    case PEN = 'PEN';  // Peruvian Sol

    // African currencies
    case ZAR = 'ZAR';  // South African Rand
    case NGN = 'NGN';  // Nigerian Naira
    case EGP = 'EGP';  // Egyptian Pound
    case KES = 'KES';  // Kenyan Shilling
    case MAD = 'MAD';  // Moroccan Dirham

    // Oceania currencies
    case FJD = 'FJD';  // Fiji Dollar

    // Precious metals and special
    case XAU = 'XAU';  // Gold (troy ounce)
    case XAG = 'XAG';  // Silver (troy ounce)
    case XPT = 'XPT';  // Platinum (troy ounce)
    case XBT = 'XBT';  // Bitcoin (unofficial)

    /**
     * Get the number of minor units (decimal places) for this currency.
     *
     * @return int Number of decimal places (0, 2, 3, or 8 for crypto)
     */
    public function minorUnits(): int
    {
        return match ($this) {
            // Zero decimal places
            self::JPY, self::KRW, self::VND, self::ISK,
            self::CLP, self::HUF, self::IDR => 0,

            // Three decimal places
            self::KWD, self::BHD, self::OMR, self::JOD => 3,

            // Precious metals (no minor units)
            self::XAU, self::XAG, self::XPT => 0,

            // Crypto (8 decimal places - satoshis)
            self::XBT => 8,

            // Standard two decimal places
            default => 2,
        };
    }

    /**
     * Get the currency symbol.
     *
     * @return string The currency symbol or code if no symbol
     */
    public function symbol(): string
    {
        return match ($this) {
            self::USD, self::AUD, self::CAD, self::NZD,
            self::HKD, self::SGD, self::MXN, self::ARS,
            self::CLP, self::COP => '$',
            self::EUR => "\u{20AC}",
            self::GBP, self::EGP => "\u{00A3}",
            self::JPY, self::CNY => "\u{00A5}",
            self::CHF => 'CHF',
            self::INR => "\u{20B9}",
            self::KRW => "\u{20A9}",
            self::RUB => "\u{20BD}",
            self::TRY => "\u{20BA}",
            self::ILS => "\u{20AA}",
            self::PLN => "z\u{0142}",
            self::BRL => 'R$',
            self::ZAR => 'R',
            self::THB => "\u{0E3F}",
            self::UAH => "\u{20B4}",
            self::VND => "\u{20AB}",
            self::PHP => "\u{20B1}",
            self::XBT => "\u{20BF}",
            default => $this->value,
        };
    }

    /**
     * Get the full currency name.
     *
     * @return string The currency name
     */
    public function name(): string
    {
        return match ($this) {
            self::USD => 'United States Dollar',
            self::EUR => 'Euro',
            self::GBP => 'British Pound Sterling',
            self::JPY => 'Japanese Yen',
            self::CHF => 'Swiss Franc',
            self::CNY => 'Chinese Yuan Renminbi',
            self::AUD => 'Australian Dollar',
            self::CAD => 'Canadian Dollar',
            self::NZD => 'New Zealand Dollar',
            self::HKD => 'Hong Kong Dollar',
            self::SGD => 'Singapore Dollar',
            self::SEK => 'Swedish Krona',
            self::NOK => 'Norwegian Krone',
            self::DKK => 'Danish Krone',
            self::PLN => 'Polish Zloty',
            self::CZK => 'Czech Koruna',
            self::HUF => 'Hungarian Forint',
            self::RON => 'Romanian Leu',
            self::BGN => 'Bulgarian Lev',
            self::ISK => 'Icelandic Krona',
            self::RUB => 'Russian Ruble',
            self::UAH => 'Ukrainian Hryvnia',
            self::TRY => 'Turkish Lira',
            self::KRW => 'South Korean Won',
            self::INR => 'Indian Rupee',
            self::IDR => 'Indonesian Rupiah',
            self::MYR => 'Malaysian Ringgit',
            self::THB => 'Thai Baht',
            self::PHP => 'Philippine Peso',
            self::VND => 'Vietnamese Dong',
            self::TWD => 'Taiwan Dollar',
            self::AED => 'UAE Dirham',
            self::SAR => 'Saudi Riyal',
            self::ILS => 'Israeli New Shekel',
            self::QAR => 'Qatari Riyal',
            self::KWD => 'Kuwaiti Dinar',
            self::BHD => 'Bahraini Dinar',
            self::OMR => 'Omani Rial',
            self::JOD => 'Jordanian Dinar',
            self::MXN => 'Mexican Peso',
            self::BRL => 'Brazilian Real',
            self::ARS => 'Argentine Peso',
            self::CLP => 'Chilean Peso',
            self::COP => 'Colombian Peso',
            self::PEN => 'Peruvian Sol',
            self::ZAR => 'South African Rand',
            self::NGN => 'Nigerian Naira',
            self::EGP => 'Egyptian Pound',
            self::KES => 'Kenyan Shilling',
            self::MAD => 'Moroccan Dirham',
            self::FJD => 'Fiji Dollar',
            self::XAU => 'Gold (troy ounce)',
            self::XAG => 'Silver (troy ounce)',
            self::XPT => 'Platinum (troy ounce)',
            self::XBT => 'Bitcoin',
        };
    }

    /**
     * Get the multiplier to convert major units to minor units.
     *
     * @return int The multiplier (e.g., 100 for USD, 1000 for KWD)
     */
    public function minorUnitMultiplier(): int
    {
        return (int)pow(10, $this->minorUnits());
    }
}

/**
 * Immutable monetary value with safe arithmetic operations.
 *
 * All amounts are stored in minor units (cents, pence, etc.) to avoid
 * floating-point precision issues. Operations return null on overflow.
 */
readonly class Money
{
    /**
     * Create a Money instance.
     *
     * @param int $minorUnits Amount in minor units (e.g., cents)
     * @param CurrencyCode $currency The currency
     */
    public function __construct(
        public int $minorUnits,
        public CurrencyCode $currency,
    ) {}

    /**
     * Create money from a major unit amount.
     *
     * @param string|int|float $amount The amount in major units
     * @param CurrencyCode $currency The currency
     * @return self|null Money instance or null if conversion fails
     */
    public static function fromMajorUnits(string|int|float $amount, CurrencyCode $currency): ?self
    {
        $multiplier = $currency->minorUnitMultiplier();

        if (is_string($amount)) {
            // Parse string amount carefully
            if (!preg_match('/^-?\d+(?:\.\d+)?$/', $amount)) {
                return null;
            }
            $floatAmount = (float)$amount;
        } else {
            $floatAmount = (float)$amount;
        }

        $minorUnits = (int)round($floatAmount * $multiplier);

        // Check for overflow
        if (abs($floatAmount * $multiplier) > PHP_INT_MAX) {
            return null;
        }

        return new self($minorUnits, $currency);
    }

    /**
     * Create money from minor units.
     *
     * @param int $minorUnits Amount in minor units
     * @param CurrencyCode $currency The currency
     * @return self Money instance
     */
    public static function fromMinorUnits(int $minorUnits, CurrencyCode $currency): self
    {
        return new self($minorUnits, $currency);
    }

    /**
     * Create zero amount in the given currency.
     *
     * @param CurrencyCode $currency The currency
     * @return self Zero money
     */
    public static function zero(CurrencyCode $currency): self
    {
        return new self(0, $currency);
    }

    /**
     * Get the amount as a float in major units.
     *
     * Note: Use this only for display; internal calculations should use minorUnits.
     *
     * @return float The amount in major units
     */
    public function toMajorUnits(): float
    {
        return $this->minorUnits / $this->currency->minorUnitMultiplier();
    }

    /**
     * Add two monetary values.
     *
     * @param self $other The amount to add
     * @return self|null Result or null on currency mismatch/overflow
     */
    public function add(self $other): ?self
    {
        if ($this->currency !== $other->currency) {
            return null;
        }

        $result = SafeMath::addChecked($this->minorUnits, $other->minorUnits);
        if ($result === null) {
            return null;
        }

        return new self($result, $this->currency);
    }

    /**
     * Subtract a monetary value.
     *
     * @param self $other The amount to subtract
     * @return self|null Result or null on currency mismatch/overflow
     */
    public function subtract(self $other): ?self
    {
        if ($this->currency !== $other->currency) {
            return null;
        }

        $result = SafeMath::subChecked($this->minorUnits, $other->minorUnits);
        if ($result === null) {
            return null;
        }

        return new self($result, $this->currency);
    }

    /**
     * Multiply by a factor.
     *
     * @param int|float $factor The multiplier
     * @param int $roundingMode PHP rounding mode constant
     * @return self|null Result or null on overflow
     */
    public function multiply(int|float $factor, int $roundingMode = PHP_ROUND_HALF_UP): ?self
    {
        $result = $this->minorUnits * $factor;

        // Check for overflow
        if (abs($result) > PHP_INT_MAX) {
            return null;
        }

        return new self((int)round($result, 0, $roundingMode), $this->currency);
    }

    /**
     * Divide by a divisor.
     *
     * @param int|float $divisor The divisor
     * @param int $roundingMode PHP rounding mode constant
     * @return self|null Result or null on division by zero
     */
    public function divide(int|float $divisor, int $roundingMode = PHP_ROUND_HALF_UP): ?self
    {
        if ($divisor == 0) {
            return null;
        }

        $result = $this->minorUnits / $divisor;
        return new self((int)round($result, 0, $roundingMode), $this->currency);
    }

    /**
     * Get the absolute value.
     *
     * @return self|null Absolute value or null on overflow (MIN_INT case)
     */
    public function abs(): ?self
    {
        $result = SafeMath::absSafe($this->minorUnits);
        if ($result === null) {
            return null;
        }
        return new self($result, $this->currency);
    }

    /**
     * Negate the amount.
     *
     * @return self|null Negated value or null on overflow
     */
    public function negate(): ?self
    {
        $result = SafeMath::mulChecked($this->minorUnits, -1);
        if ($result === null) {
            return null;
        }
        return new self($result, $this->currency);
    }

    /**
     * Check if amount is zero.
     *
     * @return bool True if zero
     */
    public function isZero(): bool
    {
        return $this->minorUnits === 0;
    }

    /**
     * Check if amount is positive.
     *
     * @return bool True if positive
     */
    public function isPositive(): bool
    {
        return $this->minorUnits > 0;
    }

    /**
     * Check if amount is negative.
     *
     * @return bool True if negative
     */
    public function isNegative(): bool
    {
        return $this->minorUnits < 0;
    }

    /**
     * Compare to another monetary value.
     *
     * @param self $other The value to compare with
     * @return int|null -1, 0, 1, or null on currency mismatch
     */
    public function compare(self $other): ?int
    {
        if ($this->currency !== $other->currency) {
            return null;
        }
        return $this->minorUnits <=> $other->minorUnits;
    }

    /**
     * Check equality with another monetary value.
     *
     * @param self $other The value to compare with
     * @return bool True if equal (same currency and amount)
     */
    public function equals(self $other): bool
    {
        return $this->currency === $other->currency
            && $this->minorUnits === $other->minorUnits;
    }

    /**
     * Check if greater than another value.
     *
     * @param self $other The value to compare with
     * @return bool|null True if greater, null on currency mismatch
     */
    public function greaterThan(self $other): ?bool
    {
        $cmp = $this->compare($other);
        return $cmp === null ? null : $cmp > 0;
    }

    /**
     * Check if less than another value.
     *
     * @param self $other The value to compare with
     * @return bool|null True if less, null on currency mismatch
     */
    public function lessThan(self $other): ?bool
    {
        $cmp = $this->compare($other);
        return $cmp === null ? null : $cmp < 0;
    }

    /**
     * Allocate money across ratios (useful for tax calculations).
     *
     * @param array<int> $ratios The ratios to allocate across
     * @return array<self>|null Array of allocated amounts or null if invalid
     */
    public function allocate(array $ratios): ?array
    {
        if (empty($ratios)) {
            return null;
        }

        $total = array_sum($ratios);
        if ($total <= 0) {
            return null;
        }

        $remainder = $this->minorUnits;
        $results = [];

        foreach ($ratios as $ratio) {
            $share = (int)floor($this->minorUnits * $ratio / $total);
            $results[] = new self($share, $this->currency);
            $remainder -= $share;
        }

        // Distribute remainder to first shares
        for ($i = 0; $remainder > 0 && $i < count($results); $i++) {
            $results[$i] = new self($results[$i]->minorUnits + 1, $this->currency);
            $remainder--;
        }

        return $results;
    }

    /**
     * Format as a string with currency symbol.
     *
     * @param bool $symbolFirst Whether symbol comes before amount
     * @return string Formatted string
     */
    public function format(bool $symbolFirst = true): string
    {
        $decimals = $this->currency->minorUnits();
        $amount = number_format($this->toMajorUnits(), $decimals, '.', ',');
        $symbol = $this->currency->symbol();

        if ($symbolFirst) {
            return $symbol . $amount;
        }
        return $amount . ' ' . $symbol;
    }

    /**
     * Format as ISO string (e.g., "USD 123.45").
     *
     * @return string ISO-formatted string
     */
    public function formatIso(): string
    {
        $decimals = $this->currency->minorUnits();
        $amount = number_format($this->toMajorUnits(), $decimals, '.', '');
        return $this->currency->value . ' ' . $amount;
    }

    /**
     * Get string representation.
     *
     * @return string Formatted money string
     */
    public function __toString(): string
    {
        return $this->format();
    }
}

/**
 * Safe currency operations and utilities.
 */
class SafeCurrency
{
    /**
     * Parse a currency code string.
     *
     * @param string $code The 3-letter currency code
     * @return CurrencyCode|null The currency or null if invalid
     */
    public static function parseCode(string $code): ?CurrencyCode
    {
        $normalized = strtoupper(trim($code));
        return CurrencyCode::tryFrom($normalized);
    }

    /**
     * Check if a currency code is valid.
     *
     * @param string $code The code to check
     * @return bool True if valid
     */
    public static function isValidCode(string $code): bool
    {
        return self::parseCode($code) !== null;
    }

    /**
     * Parse a monetary string (e.g., "USD 123.45" or "$123.45").
     *
     * @param string $value The string to parse
     * @param CurrencyCode|null $defaultCurrency Currency to use if not in string
     * @return Money|null The parsed money or null
     */
    public static function parse(string $value, ?CurrencyCode $defaultCurrency = null): ?Money
    {
        $value = trim($value);

        // Try ISO format first: "USD 123.45"
        if (preg_match('/^([A-Z]{3})\s+(-?\d+(?:\.\d+)?)$/', $value, $matches)) {
            $currency = self::parseCode($matches[1]);
            if ($currency !== null) {
                return Money::fromMajorUnits($matches[2], $currency);
            }
        }

        // Try amount only with default currency
        if ($defaultCurrency !== null && preg_match('/^-?\d+(?:\.\d+)?$/', $value)) {
            return Money::fromMajorUnits($value, $defaultCurrency);
        }

        return null;
    }

    /**
     * Get all available currency codes.
     *
     * @return array<CurrencyCode> All currency codes
     */
    public static function allCurrencies(): array
    {
        return CurrencyCode::cases();
    }

    /**
     * Get currencies with zero decimal places.
     *
     * @return array<CurrencyCode> Zero-decimal currencies
     */
    public static function zeroDecimalCurrencies(): array
    {
        return array_filter(
            CurrencyCode::cases(),
            fn(CurrencyCode $c) => $c->minorUnits() === 0
        );
    }

    /**
     * Get currencies with three decimal places.
     *
     * @return array<CurrencyCode> Three-decimal currencies
     */
    public static function threeDecimalCurrencies(): array
    {
        return array_filter(
            CurrencyCode::cases(),
            fn(CurrencyCode $c) => $c->minorUnits() === 3
        );
    }
}
